Src/Sarif.PatternMatcher/AnalyzeCommand.cs

```diff
@@ -122,36 +122,79 @@ internal static void PushInheritedData(SearchDefinitions definitions, Dictionary
+
+            definition.FileNameAllowRegex = PushData(definition.FileNameAllowRegex,
+                                                     definition.SharedStrings,
+                                                     sharedStrings);
 
             foreach (MatchExpression matchExpression in definition.MatchExpressions)
             {
-                matchExpression.FileNameDenyRegex = PushData(matchExpression.FileNameDenyRegex, sharedStrings);
+                matchExpression.FileNameDenyRegex = PushData(matchExpression.FileNameDenyRegex,
+                                                             definition.SharedStrings,
+                                                             sharedStrings);
+
                 matchExpression.FileNameDenyRegex ??= definition.FileNameDenyRegex;
 
-                matchExpression.FileNameAllowRegex = PushData(matchExpression.FileNameAllowRegex, sharedStrings);
+                matchExpression.FileNameAllowRegex = PushData(matchExpression.FileNameAllowRegex,
+                                                             definition.SharedStrings,
+                                                             sharedStrings);
+
                 matchExpression.FileNameAllowRegex ??= definition.FileNameAllowRegex;
 
-                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex, sharedStrings);
+                matchExpression.ContentsRegex = PushData(matchExpression.ContentsRegex,
+                                                         definition.SharedStrings,
+                                                         sharedStrings);
+
+                if (matchExpression.Level == 0)
+                {
+                    matchExpression.Level = definition.Level;
+                }
+            }
+        }
+
+        private static string PushData(string text, params Dictionary<string, string>[] sharedStringsDictionaries)
+        {
+            if (text?.Contains("$") != true)
+            {
+                return text;
+            }
+
+            foreach (Dictionary<string, string> sharedStrings in sharedStringsDictionaries)
+            {
+                if (sharedStrings == null)
+                {
+                    continue;
+                }
 
-                if (matchExpression.Level == 0) { matchExpression.Level = definition.Level; }
+                foreach (string key in sharedStrings.Keys)
+                {
+                    text = text.Replace(key, sharedStrings[key]);
+                }
             }
+
+            return text;
         }
 
         internal static Dictionary<string, string> LoadSharedStrings(string sharedStringsFullPath, IFileSystem fileSystem)
         {
             var result = new Dictionary<string, string>();
 
-            foreach (string line in fileSystem.FileReadAllLines(sharedStringsFullPath))
+            foreach (string fileLine in fileSystem.FileReadAllLines(sharedStringsFullPath))
```

Comment:
>FileReadAllLines [](start = 51, length = 16)

nit: we could change from ReadAllLines to ReadLines. The first adds everything to memory the other uses ienumerable, so, one line is in memory.

Reply:
good suggestion.

---
In reply to: [555679261](https://github.com/microsoft/sarif-pattern-matcher/pull/101#discussion_r555679261) [](ancestors = 555679261)